src/compiler/checker.ts

```diff
@@ -28959,6 +28959,15 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
                     return isCircularMappedProperty(prop) ? undefined : getTypeOfSymbol(prop);
                 }
                 if (isTupleType(t) && isNumericLiteralName(name) && +name >= 0) {
+                    if (t.target.hasRestElement) {
+                        const restElement = getTypeArguments(t)[t.target.fixedLength];
+                        if (restElement !== type) {
+                            const propType = getTypeOfPropertyOfContextualType(restElement, name);

```

Comment:
Erm, `name` needs to be offset by the length of the preceding fixed elements of the tuple, no? And even that's thrown into question with multiple variadic tuple elements? Eg, if `t` is `[...T, ...U]` and `name` is 0, where `T extends [] | [number]` and `U extends [] | [string]`, this is just going to lookup `0` on `T`, which is, at best, a partial result. (You'd want the contextual type to be something like `string | number | undefined` or something constrained to such). Seems like something a bit more complete is needed to handle variadic tuples fully.